benchmarks/src/test/java/org/apache/druid/timeline/DimensionRangeShardSpecBenchmark.java

```diff
@@ -0,0 +1,146 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.timeline;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableRangeSet;
+import com.google.common.collect.Range;
+import com.google.common.collect.RangeSet;
+import org.apache.druid.common.config.NullHandling;
+import org.apache.druid.data.input.StringTuple;
+import org.apache.druid.java.util.common.StringUtils;
+import org.apache.druid.query.filter.InDimFilter;
+import org.apache.druid.timeline.partition.DimensionRangeShardSpec;
+import org.junit.Assert;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OperationsPerInvocation;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+@State(Scope.Benchmark)
+@Fork(value = 1, jvmArgsAppend = {"-Xms128m", "-Xmx128m", "-XX:+UseG1GC"})
+@Warmup(iterations = 10)
+@Measurement(iterations = 10)
+@BenchmarkMode({Mode.AverageTime})
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@OperationsPerInvocation(3) // 3 shard specs per invocation
+@SuppressWarnings("UnstableApiUsage") // Range, RangeSet are unstable APIs
+public class DimensionRangeShardSpecBenchmark
+{
+  private final Map<String, RangeSet<String>> domainSinglePrunable =
+      ImmutableMap.<String, RangeSet<String>>builder()
+                  .put("country", ImmutableRangeSet.of(Range.singleton("India")))
+                  .build();
+
+  // Initialized in @Setup method.
+  private Map<String, RangeSet<String>> domain5kPrunable;
+
+  private final Map<String, RangeSet<String>> domainSingleNonPrunable =
+      ImmutableMap.<String, RangeSet<String>>builder()
+                  .put("country", ImmutableRangeSet.of(Range.singleton("Shanghai")))
+                  .build();
+
+  // Initial segment (null -> values)
+  private final DimensionRangeShardSpec shardSpec0 = new DimensionRangeShardSpec(
+      Arrays.asList("country", "city"),
+      new StringTuple(new String[]{null, null}),
```

Comment:
Nit: start of first segment is actually denoted by the whole tuple itself being null, rather than a tuple of null values. Same for end of last segment. But for the purposes of these benchmarks/tests, I guess it doesn't matter.